### PR TITLE
seed/seedwriter/writer.go: check DevModeConfinement for dangerous features

### PR DESCRIPTION
**Pull request by anonymouse64**
This will enforce that devmode snaps can only be used where dangerous features
are allowed, i.e. dangerous grade models.

Also change the generic error message we return to actually be **_even_** more
generic and apply to more error cases :-/ because code is hard.

### CODE REVIEW

File: seed/seedwriter/writer_test.go [hunk 1]
```diff
@@ -2111,17 +2119,42 @@ func (s *writerSuite) TestCore20NonDangerousNoChannelOverride(c *C) {
...
+	sn := snaps[4]
+
+	info := s.AssertedSnapInfo(sn.SnapName())
+	c.Assert(info, NotNil, Commentf("%s not defined", sn.SnapName()))
+	err = w.SetInfo(sn, info)
+	c.Assert(err, ErrorMatches, "cannot override channels, add devmode snaps, local snaps, or extra snaps with a model of grade higher than dangerous")
 }
 
-func (s *writerSuite) TestCore20NonDangerousDisallowedOptionsSnaps(c *C) {
+func (s *writerSuite) TestCore20NonDangerousNoChannelOverride(c *C) {
```
Comment (pedronis) [inline seed/seedwriter/writer_test.go:2157]: did you merge two tests here? or one went away? bit confused by the diff here

Comment (anonymouse64) [reply seed/seedwriter/writer_test.go:2157]: I don't know why the diff is so confusing but I just added one test I didn't modify the other one but git thinks I did something much more complicated

Comment (pedronis) [reply seed/seedwriter/writer_test.go:2157]: no,  view file seems to agree that TestCore20NonDangerousDisallowedOptionsSnaps has disappeared

**Comment by pedronis** [comment]
a test seems to have gone missing here

File: seed/seedwriter/writer_test.go (hunk 1 above)
Comment (anonymouse64) [reply seed/seedwriter/writer_test.go:2234]: sorry not sure how that happened, I have restore the test with a force push

File: seed/seedwriter/writer.go [hunk 2]
```diff
@@ -547,6 +547,12 @@ func (w *Writer) SetInfo(sn *SeedSnap, info *snap.Info) error {
 		return err
 	}
 
+	if info.Confinement == snap.DevModeConfinement {
+		if err := w.policy.allowsDangerousFeatures(); err != nil {
```
Comment (pedronis) [inline seed/seedwriter/writer.go:551]: to improve the error, to put in focus what's the problem is, we could pass a feature in play enum value to this function. Doesn't seem something to do now though

**Review by pedronis — APPROVED** [review]
thanks

File: seed/seedwriter/writer.go (hunk 2 above)
Comment (anonymouse64) [reply seed/seedwriter/writer.go:551]: yeah as annoying as it is for me personally, most people won't see this unless they see it from ubuntu-image, and even then they will be building their own images which is an infrequent task I think.

File: seed/seedwriter/writer.go [hunk 3]
```diff
@@ -547,6 +547,12 @@ func (w *Writer) SetInfo(sn *SeedSnap, info *snap.Info) error {
 		return err
 	}
 
+	if info.Confinement == snap.DevModeConfinement {
```
[the remark this answers is not in the document]
Comment (anonymouse64) [reply seed/seedwriter/writer.go:550]: yes for prepare-image this was the best place that I could find, but for the more general problem of enabling devmode snaps only on grade dangerous (which will be a separate pr and maybe not in time for 1.0) it's very unclear where the right spots are to handle all the different corner cases, but we can sort that out when we get there

[the remark this answers is not in the document]
Comment (zyga) [reply seed/seedwriter/writer.go:550]: Should we check earlier? Perhaps before setting `sn.Info`? Right now this has the odd property of having both side effects and errors.

File: seed/seedwriter/writer_test.go [hunk 4]
```diff
@@ -2111,14 +2119,39 @@ func (s *writerSuite) TestCore20NonDangerousNoChannelOverride(c *C) {
...
+	c.Check(err, IsNil)
+	c.Assert(snaps, HasLen, 5)
+
+	c.Assert(snaps[4].SnapName(), Equals, "my-devmode")
+	sn := snaps[4]
+
+	info := s.AssertedSnapInfo(sn.SnapName())
+	c.Assert(info, NotNil, Commentf("%s not defined", sn.SnapName()))
+	err = w.SetInfo(sn, info)
+	c.Assert(err, ErrorMatches, "cannot override channels, add devmode snaps, local snaps, or extra snaps with a model of grade higher than dangerous")
```
Comment (zyga) [inline seed/seedwriter/writer_test.go:2154]: An extra check that sn is not modified would be good, perhaps.

**Review by zyga — APPROVED** [review]
+1 with a comment inline

File: seed/seedwriter/writer_test.go (hunk 4 above)
Comment (zyga) [reply seed/seedwriter/writer_test.go:2154]: Pushed myself.

**Review by bboozzoo — APPROVED** [review]
LGTM

**Pull request merged**